Error lines in https://jenkins.akka.io/job/pr-validator-per-commit-jenkins/3328/:

```
 10312 Test akka.stream.javadsl.FlowTest.mustBeAbleToRecover failed: java.lang.AssertionError: assertion failed: timeout (3 seconds) during expectMsg while waiting for 1, took 3.025 sec
 10313     at scala.Predef$.assert(Predef.scala:179)
 10314     at akka.testkit.TestKitBase$class.expectMsg_internal(TestKit.scala:338)
 10315     at akka.testkit.TestKitBase$class.expectMsg(TestKit.scala:315)
 10316     at akka.testkit.TestKit.expectMsg(TestKit.scala:718)
 10317     at akka.testkit.JavaTestKit.expectMsgEquals(JavaTestKit.java:378)
 10318     at akka.stream.javadsl.FlowTest.mustBeAbleToRecover(FlowTest.java:548)
 10319     ...
 12511 Failed: Total 867, Failed 1, Errors 0, Passed 866, Pending 22
 12512 Failed tests:
 12513 	akka.stream.javadsl.FlowTest
 15887 (akka-stream-tests-experimental/test:test) sbt.TestsFailedException: Tests unsuccessful
 15888 Total time: 858 s, completed Aug 7, 2015 4:34:39 PM
```

--- a/akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala
+++ b/akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala
@@ -298,11 +298,7 @@ private[http] object HttpServerBluePrint {
     val sink = StreamUtils.oneTimePublisherSink[FrameEvent](sinkCell, "frameHandler.in")
     val source = StreamUtils.oneTimeSubscriberSource[FrameEvent](sourceCell, "frameHandler.out")
 
-    val flow =
-      Flow[ByteString]
-        .transform[FrameEvent](() ⇒ new FrameEventParser)
-        .via(Flow.wrap(sink, source)(Keep.none))
-        .transform(() ⇒ new FrameEventRenderer)
+    val flow = Websocket.framing.join(Flow.wrap(sink, source)(Keep.none))
 
     new WebsocketSetup {
       def websocketFlow: Flow[ByteString, ByteString, Any] = flow

--- a/akka-http-core/src/main/scala/akka/http/impl/engine/ws/FrameHandler.scala
+++ b/akka-http-core/src/main/scala/akka/http/impl/engine/ws/FrameHandler.scala
@@ -18,11 +18,12 @@ import scala.util.control.NonFatal
  * INTERNAL API
  */
 private[http] object FrameHandler {
-  def create(server: Boolean): Flow[FrameEvent, Either[BypassEvent, MessagePart], Unit] =
+  type Output = Either[BypassEvent, MessagePart]
+  def create(server: Boolean): Flow[FrameEvent, Output, Unit] =
     Flow[FrameEvent].transform(() ⇒ new HandlerStage(server))
 
-  class HandlerStage(server: Boolean) extends StatefulStage[FrameEvent, Either[BypassEvent, MessagePart]] {
-    type Ctx = Context[Either[BypassEvent, MessagePart]]
+  class HandlerStage(server: Boolean) extends StatefulStage[FrameEvent, Output] {
+    type Ctx = Context[Output]
     def initial: State = Idle
 
     object Idle extends StateWithControlFrameHandling {
@@ -129,7 +130,7 @@ private[http] object FrameHandler {
           Right(ActivelyCloseWithCode(Some(closeCode), reason))), ctx, CloseAfterPeerClosed)
 
     object CloseAfterPeerClosed extends State {
-      def onPush(elem: FrameEvent, ctx: Context[Either[BypassEvent, MessagePart]]): SyncDirective =
+      def onPush(elem: FrameEvent, ctx: Context[Output]): SyncDirective =
         elem match {
           case FrameStart(FrameHeader(Opcode.Close, _, length, _, _, _, _), data) ⇒
             become(WaitForPeerTcpClose)
@@ -139,7 +140,7 @@ private[http] object FrameHandler {
         }
     }
     object WaitForPeerTcpClose extends State {
-      def onPush(elem: FrameEvent, ctx: Context[Either[BypassEvent, MessagePart]]): SyncDirective =
+      def onPush(elem: FrameEvent, ctx: Context[Output]): SyncDirective =
         ctx.pull() // ignore
     }
 

--- a/akka-http-core/src/main/scala/akka/http/impl/engine/ws/FrameOutHandler.scala
+++ b/akka-http-core/src/main/scala/akka/http/impl/engine/ws/FrameOutHandler.scala
@@ -4,6 +4,8 @@
 
 package akka.http.impl.engine.ws
 
+import akka.stream.scaladsl.Flow
+
 import scala.concurrent.duration.FiniteDuration
 
 import akka.stream.stage._
@@ -16,7 +18,7 @@ import Websocket.Tick
  *
  * INTERNAL API
  */
-private[http] class FrameOutHandler(serverSide: Boolean, _closeTimeout: FiniteDuration) extends StatefulStage[AnyRef, FrameStart] {
+private[http] class FrameOutHandler(serverSide: Boolean, _closeTimeout: FiniteDuration) extends StatefulStage[FrameOutHandler.Input, FrameStart] {
   def initial: StageState[AnyRef, FrameStart] = Idle
   def closeTimeout: Timestamp = Timestamp.now + _closeTimeout
 
@@ -129,4 +131,11 @@ private[http] class FrameOutHandler(serverSide: Boolean, _closeTimeout: FiniteDu
       ctx.absorbTermination()
     case _ ⇒ super.onUpstreamFailure(cause, ctx)
   }
+}
+
+private[http] object FrameOutHandler {
+  type Input = AnyRef
+
+  def create(serverSide: Boolean, closeTimeout: FiniteDuration): Flow[Input, FrameStart, Unit] =
+    Flow[Input].transform(() ⇒ new FrameOutHandler(serverSide, closeTimeout))
 }

--- a/akka-http-core/src/main/scala/akka/http/impl/engine/ws/Masking.scala
+++ b/akka-http-core/src/main/scala/akka/http/impl/engine/ws/Masking.scala
@@ -4,7 +4,7 @@
 
 package akka.http.impl.engine.ws
 
-import akka.stream.scaladsl.Flow
+import akka.stream.scaladsl.{ Keep, BidiFlow, Flow }
 import akka.stream.stage.{ SyncDirective, Context, StageState, StatefulStage }
 
 import scala.util.Random
@@ -15,6 +15,9 @@ import scala.util.Random
  * INTERNAL API
  */
 private[http] object Masking {
+  def apply(serverSide: Boolean, maskRandom: () ⇒ Random): BidiFlow[ /* net in */ FrameEvent, /* app out */ FrameEvent, /* app in */ FrameEvent, /* net out */ FrameEvent, Unit] =
+    BidiFlow.wrap(unmaskIf(serverSide), maskIf(!serverSide, maskRandom))(Keep.none)
+
   def maskIf(condition: Boolean, maskRandom: () ⇒ Random): Flow[FrameEvent, FrameEvent, Unit] =
     if (condition) Flow[FrameEvent].transform(() ⇒ new Masking(maskRandom())) // new random per materialization
     else Flow[FrameEvent]

--- a/akka-http-core/src/main/scala/akka/http/impl/engine/ws/UpgradeToWebsocketLowLevel.scala
+++ b/akka-http-core/src/main/scala/akka/http/impl/engine/ws/UpgradeToWebsocketLowLevel.scala
@@ -28,5 +28,5 @@ private[http] abstract class UpgradeToWebsocketLowLevel extends InternalCustomHe
   private[http] def handleFrames(handlerFlow: Flow[FrameEvent, FrameEvent, Any], subprotocol: Option[String] = None): HttpResponse
 
   override def handleMessages(handlerFlow: Flow[Message, Message, Any], subprotocol: Option[String] = None): HttpResponse =
-    handleFrames(Websocket.handleMessages(handlerFlow), subprotocol)
+    handleFrames(Websocket.stack(serverSide = true).join(handlerFlow), subprotocol)
 }

--- a/akka-http-core/src/main/scala/akka/http/impl/engine/ws/Websocket.scala
+++ b/akka-http-core/src/main/scala/akka/http/impl/engine/ws/Websocket.scala
@@ -6,9 +6,11 @@ package akka.http.impl.engine.ws
 
 import java.security.SecureRandom
 
+import akka.util.ByteString
+
 import scala.concurrent.duration._
 
-import akka.stream.{ Attributes, FanOutShape2, FanInShape3, Inlet }
+import akka.stream._
 import akka.stream.scaladsl._
 import akka.stream.stage._
 import FlexiRoute.{ DemandFrom, DemandFromAny, RouteLogic }
@@ -19,13 +21,40 @@ import akka.http.scaladsl.model.ws._
 
 /**
  * INTERNAL API
+ *
+ * Defines components of the websocket stack.
  */
 private[http] object Websocket {
   import FrameHandler._
 
-  def handleMessages[T](messageHandler: Flow[Message, Message, T],
-                        serverSide: Boolean = true,
-                        closeTimeout: FiniteDuration = 3.seconds): Flow[FrameEvent, FrameEvent, Unit] = {
+  /** The lowest layer that implements the binary protocol */
+  def framing: BidiFlow[ByteString, FrameEvent, FrameEvent, ByteString, Unit] =
+    BidiFlow.wrap(
+      Flow[ByteString].transform(() ⇒ new FrameEventParser),
+      Flow[FrameEvent].transform(() ⇒ new FrameEventRenderer))(Keep.none)
+      .named("ws-framing")
+
+  /** The layer that handles masking using the rules defined in the specification */
+  def masking(serverSide: Boolean): BidiFlow[FrameEvent, FrameEvent, FrameEvent, FrameEvent, Unit] =
+    Masking(serverSide, () ⇒ new SecureRandom())
+      .named("ws-masking")
+
+  /**
+   * The layer that implements all low-level frame handling, like handling control frames, collecting messages
+   * from frames, decoding text messages, close handling, etc.
+   */
+  def frameHandling(serverSide: Boolean = true,
+                    closeTimeout: FiniteDuration): BidiFlow[FrameEvent, FrameHandler.Output, FrameOutHandler.Input, FrameStart, Unit] =
+    BidiFlow.wrap(
+      FrameHandler.create(server = serverSide),
+      FrameOutHandler.create(serverSide, closeTimeout))(Keep.none)
+      .named("ws-frame-handling")
+
+  /**
+   * The layer that provides the high-level user facing API on top of frame handling.
+   */
+  def messageAPI(serverSide: Boolean,
+                 closeTimeout: FiniteDuration): BidiFlow[FrameHandler.Output, Message, Message, FrameOutHandler.Input, Unit] = {
     /** Completes this branch of the flow if no more messages are expected and converts close codes into errors */
     class PrepareForUserHandler extends PushStage[MessagePart, MessagePart] {
       var inMessage = false
@@ -91,18 +120,6 @@ private[http] object Websocket {
       }
     }
 
-    lazy val userFlow =
-      Flow[MessagePart]
-        .transform(() ⇒ new PrepareForUserHandler)
-        .splitWhen(_.isMessageEnd) // FIXME using splitAfter from #16885 would simplify protocol a lot
-        .map(_.collect {
-          case m: MessageDataPart ⇒ m
-        })
-        .via(collectMessage)
-        .via(messageHandler)
-        .via(MessageToFrameRenderer.create(serverSide))
-        .transform(() ⇒ new LiftCompletions)
-
     /**
      * Distributes output from the FrameHandler into bypass and userFlow.
      */
@@ -158,26 +175,55 @@ private[http] object Websocket {
         }
     }
 
-    lazy val bypassAndUserHandler: Flow[Either[BypassEvent, MessagePart], AnyRef, Unit] =
-      Flow(BypassRouter, Source(closeTimeout, closeTimeout, Tick), BypassMerge)((_, _, _) ⇒ ()) { implicit b ⇒
-        (split, tick, merge) ⇒
-          import FlowGraph.Implicits._
+    def prepareMessages: Flow[MessagePart, Message, Unit] =
+      Flow[MessagePart]
+        .transform(() ⇒ new PrepareForUserHandler)
+        .splitWhen(_.isMessageEnd) // FIXME using splitAfter from #16885 would simplify protocol a lot
+        .map(_.collect {
+          case m: MessageDataPart ⇒ m
+        })
+        .via(collectMessage)
+        .named("ws-prepare-messages")
 
-          split.out0 ~> merge.in0
-          split.out1 ~> userFlow ~> merge.in1
-          tick.outlet ~> merge.in2
+    def renderMessages: Flow[Message, FrameStart, Unit] =
+      MessageToFrameRenderer.create(serverSide)
+        .named("ws-render-messages")
 
-          (split.in, merge.out)
-      }
+    BidiFlow() { implicit b ⇒
+      import FlowGraph.Implicits._
 
-    Flow[FrameEvent]
-      .via(Masking.unmaskIf(serverSide))
-      .via(FrameHandler.create(server = serverSide))
-      .mapConcat(x ⇒ x :: x :: Nil) // FIXME: #17004
-      .via(bypassAndUserHandler)
-      .transform(() ⇒ new FrameOutHandler(serverSide, closeTimeout))
-      .via(Masking.maskIf(!serverSide, () ⇒ new SecureRandom()))
+      val routePreparation = b.add(Flow[FrameHandler.Output].mapConcat(x ⇒ x :: x :: Nil))
+      val split = b.add(BypassRouter)
+      val tick = Source(closeTimeout, closeTimeout, Tick)
+      val merge = b.add(BypassMerge)
+      val messagePreparation = b.add(prepareMessages)
+      val messageRendering = b.add(renderMessages.transform(() ⇒ new LiftCompletions))
+
+      routePreparation.outlet ~> split.in
+
+      // user handler
+      split.out1 ~> messagePreparation
+      messageRendering.outlet ~> merge.in1
+
+      // bypass
+      split.out0 ~> merge.in0
+
+      // timeout support
+      tick ~> merge.in2
+
+      BidiShape(
+        routePreparation.inlet,
+        messagePreparation.outlet,
+        messageRendering.inlet,
+        merge.out)
+    }.named("ws-message-api")
   }
+
+  def stack(serverSide: Boolean = true,
+            closeTimeout: FiniteDuration = 3.seconds): BidiFlow[FrameEvent, Message, Message, FrameEvent, Unit] =
+    masking(serverSide) atop
+      frameHandling(serverSide, closeTimeout) atop
+      messageAPI(serverSide, closeTimeout)
 
   object Tick
   case object SwitchToWebsocketToken

--- a/akka-http-core/src/main/scala/akka/http/javadsl/Http.scala
+++ b/akka-http-core/src/main/scala/akka/http/javadsl/Http.scala
@@ -113,9 +113,8 @@ class Http(system: ExtendedActorSystem) extends akka.actor.Extension {
    * Convenience method which starts a new HTTP server at the given endpoint and uses the given ``handler``
    * [[Flow]] for processing all incoming connections.
    *
-   * Note that there is no backpressure being applied to the `connections` [[Source]], i.e. all
-   * connections are being accepted at maximum rate, which, depending on the applications, might
-   * present a DoS risk!
+   * The number of concurrently accepted connections can be configured by overriding
+   * the `akka.http.server.max-connections` setting.
    */
   def bindAndHandle(handler: Flow[HttpRequest, HttpResponse, _],
                     interface: String, port: Int,
@@ -128,9 +127,8 @@ class Http(system: ExtendedActorSystem) extends akka.actor.Extension {
    * Convenience method which starts a new HTTP server at the given endpoint and uses the given ``handler``
    * [[Flow]] for processing all incoming connections.
    *
-   * Note that there is no backpressure being applied to the `connections` [[Source]], i.e. all
-   * connections are being accepted at maximum rate, which, depending on the applications, might
-   * present a DoS risk!
+   * The number of concurrently accepted connections can be configured by overriding
+   * the `akka.http.server.max-connections` setting.
    */
   def bindAndHandle(handler: Flow[HttpRequest, HttpResponse, _],
                     interface: String, port: Int,
@@ -146,9 +144,8 @@ class Http(system: ExtendedActorSystem) extends akka.actor.Extension {
    * Convenience method which starts a new HTTP server at the given endpoint and uses the given ``handler``
    * [[Flow]] for processing all incoming connections.
    *
-   * Note that there is no backpressure being applied to the `connections` [[Source]], i.e. all
-   * connections are being accepted at maximum rate, which, depending on the applications, might
-   * present a DoS risk!
+   * The number of concurrently accepted connections can be configured by overriding
+   * the `akka.http.server.max-connections` setting.
    */
   def bindAndHandleSync(handler: Function[HttpRequest, HttpResponse],
                         interface: String, port: Int,
@@ -160,9 +157,8 @@ class Http(system: ExtendedActorSystem) extends akka.actor.Extension {
    * Convenience method which starts a new HTTP server at the given endpoint and uses the given ``handler``
    * [[Flow]] for processing all incoming connections.
    *
-   * Note that there is no backpressure being applied to the `connections` [[Source]], i.e. all
-   * connections are being accepted at maximum rate, which, depending on the applications, might
-   * present a DoS risk!
+   * The number of concurrently accepted connections can be configured by overriding
+   * the `akka.http.server.max-connections` setting.
    */
   def bindAndHandleSync(handler: Function[HttpRequest, HttpResponse],
                         interface: String, port: Int,
@@ -178,9 +174,8 @@ class Http(system: ExtendedActorSystem) extends akka.actor.Extension {
    * Convenience method which starts a new HTTP server at the given endpoint and uses the given ``handler``
    * [[Flow]] for processing all incoming connections.
    *
-   * Note that there is no backpressure being applied to the `connections` [[Source]], i.e. all
-   * connections are being accepted at maximum rate, which, depending on the applications, might
-   * present a DoS risk!
+   * The number of concurrently accepted connections can be configured by overriding
+   * the `akka.http.server.max-connections` setting.
    */
   def bindAndHandleAsync(handler: Function[HttpRequest, Future[HttpResponse]],
                          interface: String, port: Int,
@@ -192,9 +187,8 @@ class Http(system: ExtendedActorSystem) extends akka.actor.Extension {
    * Convenience method which starts a new HTTP server at the given endpoint and uses the given ``handler``
    * [[Flow]] for processing all incoming connections.
    *
-   * Note that there is no backpressure being applied to the `connections` [[Source]], i.e. all
-   * connections are being accepted at maximum rate, which, depending on the applications, might
-   * present a DoS risk!
+   * The number of concurrently accepted connections can be configured by overriding
+   * the `akka.http.server.max-connections` setting.
    */
   def bindAndHandleAsync(handler: Function[HttpRequest, Future[HttpResponse]],
                          interface: String, port: Int,

--- a/akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala
+++ b/akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala
@@ -74,9 +74,8 @@ class HttpExt(config: Config)(implicit system: ActorSystem) extends akka.actor.E
    * Convenience method which starts a new HTTP server at the given endpoint and uses the given ``handler``
    * [[Flow]] for processing all incoming connections.
    *
-   * Note that there is no backpressure being applied to the `connections` [[Source]], i.e. all
-   * connections are being accepted at maximum rate, which, depending on the applications, might
-   * present a DoS risk!
+   * The number of concurrently accepted connections can be configured by overriding
+   * the `akka.http.server.max-connections` setting.
    */
   def bindAndHandle(handler: Flow[HttpRequest, HttpResponse, Any],
                     interface: String, port: Int = -1,
@@ -113,9 +112,8 @@ class HttpExt(config: Config)(implicit system: ActorSystem) extends akka.actor.E
    * Convenience method which starts a new HTTP server at the given endpoint and uses the given ``handler``
    * [[Flow]] for processing all incoming connections.
    *
-   * Note that there is no backpressure being applied to the `connections` [[Source]], i.e. all
-   * connections are being accepted at maximum rate, which, depending on the applications, might
-   * present a DoS risk!
+   * The number of concurrently accepted connections can be configured by overriding
+   * the `akka.http.server.max-connections` setting.
    */
   def bindAndHandleSync(handler: HttpRequest ⇒ HttpResponse,
                         interface: String, port: Int = -1,
@@ -128,9 +126,8 @@ class HttpExt(config: Config)(implicit system: ActorSystem) extends akka.actor.E
    * Convenience method which starts a new HTTP server at the given endpoint and uses the given ``handler``
    * [[Flow]] for processing all incoming connections.
    *
-   * Note that there is no backpressure being applied to the `connections` [[Source]], i.e. all
-   * connections are being accepted at maximum rate, which, depending on the applications, might
-   * present a DoS risk!
+   * The number of concurrently accepted connections can be configured by overriding
+   * the `akka.http.server.max-connections` setting.
    */
   def bindAndHandleAsync(handler: HttpRequest ⇒ Future[HttpResponse],
                          interface: String, port: Int = -1,

--- a/akka-http-core/src/test/scala/akka/http/impl/engine/ws/MessageSpec.scala
+++ b/akka-http-core/src/test/scala/akka/http/impl/engine/ws/MessageSpec.scala
@@ -782,7 +782,7 @@ class MessageSpec extends FreeSpec with Matchers with WithMaterializerSpec {
     Source(netIn)
       .via(printEvent("netIn"))
       .transform(() ⇒ new FrameEventParser)
-      .via(Websocket.handleMessages(messageHandler, serverSide, closeTimeout = closeTimeout))
+      .via(Websocket.stack(serverSide, closeTimeout = closeTimeout).join(messageHandler))
       .via(printEvent("frameRendererIn"))
       .transform(() ⇒ new FrameEventRenderer)
       .via(printEvent("frameRendererOut"))

--- a/akka-http/src/main/scala/akka/http/scaladsl/server/directives/FormFieldDirectives.scala
+++ b/akka-http/src/main/scala/akka/http/scaladsl/server/directives/FormFieldDirectives.scala
@@ -18,12 +18,22 @@ trait FormFieldDirectives extends ToNameReceptacleEnhancements {
   /**
    * Extracts an HTTP form field from the request.
    * Rejects the request if the defined form field matcher(s) don't match.
+   *
+   * Due to a bug in Scala 2.10, invocations of this method sometimes fail to compile with an
+   * "too many arguments for method formField" or "type mismatch" error.
+   *
+   * As a workaround add an `import FormFieldDirectives.FieldMagnet` or use Scala 2.11.x.
    */
   def formField(pdm: FieldMagnet): pdm.Out = pdm()
 
   /**
    * Extracts a number of HTTP form field from the request.
    * Rejects the request if the defined form field matcher(s) don't match.
+   *
+   * Due to a bug in Scala 2.10, invocations of this method sometimes fail to compile with an
+   * "too many arguments for method formFields" or "type mismatch" error.
+   *
+   * As a workaround add an `import FormFieldDirectives.FieldMagnet` or use Scala 2.11.x.
    */
   def formFields(pdm: FieldMagnet): pdm.Out = pdm()
 

--- a/akka-http/src/main/scala/akka/http/scaladsl/server/directives/ParameterDirectives.scala
+++ b/akka-http/src/main/scala/akka/http/scaladsl/server/directives/ParameterDirectives.scala
@@ -32,12 +32,22 @@ trait ParameterDirectives extends ToNameReceptacleEnhancements {
   /**
    * Extracts a query parameter value from the request.
    * Rejects the request if the defined query parameter matcher(s) don't match.
+   *
+   * Due to a bug in Scala 2.10, invocations of this method sometimes fail to compile with an
+   * "too many arguments for method parameter" or "type mismatch" error.
+   *
+   * As a workaround add an `import ParameterDirectives.ParamMagnet` or use Scala 2.11.x.
    */
   def parameter(pdm: ParamMagnet): pdm.Out = pdm()
 
   /**
    * Extracts a number of query parameter values from the request.
    * Rejects the request if the defined query parameter matcher(s) don't match.
+   *
+   * Due to a bug in Scala 2.10, invocations of this method sometimes fail to compile with an
+   * "too many arguments for method parameters" or "type mismatch" error.
+   *
+   * As a workaround add an `import ParameterDirectives.ParamMagnet` or use Scala 2.11.x.
    */
   def parameters(pdm: ParamMagnet): pdm.Out = pdm()
 

